zkevm-circuits/src/exp_circuit.rs

```diff
@@ -382,37 +401,137 @@ impl<F: Field> ExpCircuitConfig<F> {
+                region.name_column(|| "mul_gadget.col1", self.mul_gadget.col1);
+                region.name_column(|| "mul_gadget.col2", self.mul_gadget.col2);
+                region.name_column(|| "mul_gadget.col3", self.mul_gadget.col3);
+                region.name_column(|| "mul_gadget.col4", self.mul_gadget.col4);
+                // region.name_column(|| "mul_gadget.overflow", self.mul_gadget.overflow);
+                region.name_column(|| "parity_check.col0", self.parity_check.col0);
+                region.name_column(|| "parity_check.col1", self.parity_check.col1);
+                region.name_column(|| "parity_check.col2", self.parity_check.col2);
+                region.name_column(|| "parity_check.col3", self.parity_check.col3);
+                region.name_column(|| "parity_check.col4", self.parity_check.col4);
+                // region.name_column(|| "parity_check.overflow", self.parity_check.overflow);
+                debug_assert!(
+                    max_exp_rows >= self.minimum_rows,
+                    "max_exp_rows: {}, minimum_rows: {}",
+                    max_exp_rows,
+                    self.minimum_rows,
+                );
+                let mut offset = 0;
+                for exp_event in exp_events.iter() {
+                    self.assign_exp_event(
+                        &mut region,
+                        &mut offset,
+                        exp_event,
+                        &mut mul_chip,
+                        &mut parity_check_chip,
+                    )?;
+                }
+                assert!(offset <= max_exp_rows - UNUSABLE_EXP_ROWS);
+                dbg!(max_exp_rows);
+                dbg!(self.minimum_rows);
+                // Fill the rest of the circuit with valid rows to achieve a constant assignment
+                // to the q_usable fixed column.
+                let (one, two) = (U256::one(), U256::from(2u64));
+                let pad_exp_event = ExpEvent {
+                    identifier: 0,
+                    base: one,
+                    exponent: two,
+                    exponentiation: one,
+                    steps: vec![ExpStep {
+                        a: one,
+                        b: one,
+                        d: one,
+                    }],
+                };
+                while offset + OFFSET_INCREMENT <= max_exp_rows - UNUSABLE_EXP_ROWS {
+                    self.assign_exp_event(
+                        &mut region,
+                        &mut offset,
+                        &pad_exp_event,
+                        &mut mul_chip,
+                        &mut parity_check_chip,
+                    )?;
+                }
+
+                // Fill extra unused rows required by the rotations at the last `q_enable`.
+                self.assign_unused_rows(&mut region, offset)?;
+                Ok(())
+            },
+        )
+    }
+
+    fn min_num_rows(exp_events: &[ExpEvent]) -> usize {
+        exp_events
```

Comment:
this should also account for the `unusable` rows that you'll append after assigning all steps. So:
```
exp_events.iter().map(|e| e.steps.len() * OFFSET_INCREMENT).sum() + UNUSABLE_EXP_ROWS
```

However I am not sure how/where the `min_num_rows_block` is used in our codebase.

@ed255 can may be guide you on this one, whether or not to include the unusable rows at the end also?

Reply:
> However I am not sure how/where the `min_num_rows_block` is used in our codebase.

I checked it and apparently this method is not used within our repository (maybe it's used externally?) but it's still useful to find out the circuit size for different blocks, which helps determine the `k` for a real scenario.

> @ed255 can may be guide you on this one, whether or not to include the unusable rows at the end also?

I think unusable rows should be included!  They are rows required for the circuit to work (at least for the MockProver verification; I'm not sure a bout the real prover, but even then it's just a tiny amount of rows).

Reply:
> However I am not sure how/where the `min_num_rows_block` is used in our codebase.

I checked it and apparently this method is not used within our repository (maybe it's used externally?) but it's still useful to find out the circuit size for different blocks, which helps determine the `k` for a real scenario.

> @ed255 can may be guide you on this one, whether or not to include the unusable rows at the end also?

I think unusable rows should be included!  They are rows required for the circuit to work (at least for the MockProver verification; I'm not sure a bout the real prover, but even then it's just a tiny amount of rows).